src/phoenix/experimental/evals/models/openai.py

```diff
@@ -247,15 +235,31 @@ def get_token_count_from_messages(self, messages: List[Dict[str, str]]) -> int:
             tokens_per_message = 3
             tokens_per_name = 1
 
-        encoding = self.tiktoken_encoding
-
         token_count = 0
         for message in messages:
             token_count += tokens_per_message
-            for key, value in message.items():
-                token_count += len(encoding.encode(value))
+            for key, text in message.items():
+                token_count += self.get_token_count_from_text(text)
                 if key == "name":
                     token_count += tokens_per_name
         # every reply is primed with <|start|>assistant<|message|>
         token_count += 3
         return token_count
+
+    def get_token_count_from_text(self, text: str) -> int:
+        return len(self.encoder.encode(text))
+
+    def get_tokens_from_text(self, text: str) -> List[int]:
+        return self.encoder.encode(text)
+
+    def get_text_from_tokens(self, tokens: List[int]) -> str:
+        return self.encoder.decode(tokens)
+
+
+def truncate_function(model: BaseEvalModel, text: str, token_buffer: int = 20) -> str:
+    max_token_count = model.max_context_size - token_buffer
+    if model.get_token_count_from_text(text) > max_token_count:
+        return (
+            model.get_text_from_tokens(model.get_tokens_from_text(text)[:max_token_count]) + "..."
+        )
+    return text
```

Comment:
Suggestion: We are invoking the encoder twice, once to get the token count and a second time to get the tokens themselves. We could do this just once.

Reply:
I'm confused, the encoder is simply stored in a field of the class. I don't see how that would cause a reasonable move in performance. I do see your point though. I had it different in the first pass, but I chose to create the helper methods because I can see helper functions using them.